contracts/contracts/SiweParser.sol

```diff
@@ -0,0 +1,252 @@
+     * @return VALUE and new i, if NAME matched; otherwise empty value and old i.
+     */
+    function _parseField(
+        bytes calldata str,
+        string memory name,
+        uint256 i
+    ) internal pure returns (bytes memory, uint256) {
+        uint256 j = i;
+        for (; j < str.length; j++) {
+            if (str[j] == ":") {
+                // Delimiter found, check the name.
+                if (keccak256(_substr(str, i, j)) != keccak256(bytes(name))) {
+                    return ("", i);
+                }
+
+                // Skip :
+                j++;
+                if (j < str.length && str[j] == " ") {
+                    // Skip blank
+                    j++;
+                }
+
+                i = j;
+                break;
+            }
+        }
+
+        for (; j < str.length; j++) {
+            if (str[j] == 0x0a) {
+                return (_substr(str, i, j), j + 1);
+            }
+        }
+        return (_substr(str, i, j), j);
+    }
+
+    /**
+     * @notice Parse bullets, one per line in str starting at i.
+     * @return Array of parsed values and a new i.
+     */
+    function _parseArray(bytes calldata str, uint256 i)
+        internal
+        pure
+        returns (bytes[] memory, uint256)
+    {
+        // First count the number of resources.
+        uint256 j = i;
+        uint256 count = 0;
+        for (; j < str.length - 1; j++) {
+            if (str[j] == "-" && str[j + 1] == " ") {
+                j += 2;
+                count++;
+            } else {
+                break;
+            }
+            while (j < str.length && str[j] != 0x0a) {
+                j++;
+            }
+        }
+
+        // Then build an array.
+        bytes[] memory values = new bytes[](count);
+        j = i;
+        for (uint256 c = 0; j < str.length - 1 && c != count; j++) {
+            if (str[j] == "-" && str[j + 1] == " ") {
+                i = j + 2;
+            }
+            while (j < str.length && str[j] != 0x0a) {
+                j++;
+            }
+            values[c] = _substr(str, i, j);
+            c++;
+            if (j == str.length) {
+                j--; // Subtract 1 because of the outer loop.
+            }
+        }
+        return (values, j);
+    }
+
+    /**
+     * @notice Parse SIWE message.
+     * @return ParsedSiweMessage struct with populated fields from the message.
+     */
+    function parseSiweMsg(bytes calldata siweMsg)
+        internal
+        pure
+        returns (ParsedSiweMessage memory)
+    {
+        ParsedSiweMessage memory p;
+        uint256 i = 0;
+
+        // dApp Domain.
+        for (; i < siweMsg.length; i++) {
+            if (siweMsg[i] == " ") {
+                p.schemeDomain = _substr(siweMsg, 0, i);
+                break;
+            }
+        }
+
+        i += 50; // " wants you to sign in with your Ethereum account:\n"
```

Comment:
Wow. They really thought this out.

Reply:
Yeah, the standard wants to keep a human-readable message in case the wallet doesn't support any fancy UIs around SIWE.